day4/part1.go

```diff
@@ -0,0 +1,261 @@
+    1 12 20 15 19        14 21 16 12  6         2  0 12  3  7
+
+At this point, the third board wins because it has at least one complete row or
+column of marked numbers (in this case, the entire top row is marked: 14 21 17
+24 4).
+
+The score of the winning board can now be calculated. Start by finding the sum
+of all unmarked numbers on that board; in this case, the sum is 188. Then,
+multiply that sum by the number that was just called when the board won, 24, to
+get the final score, 188 * 24 = 4512.
+
+To guarantee victory against the giant squid, figure out which board will win
+first. What will your final score be if you choose that board?
+*/
+func Part1(path string) (int, error) {
+	bingo, err := parseBingo(path)
+	if err != nil {
+		return 0, err
+	}
+
+	for _, n := range bingo.numbers {
+		for _, b := range bingo.boards {
+			if b.mark(n) {
+				fmt.Println(b)
+				return n * b.score(), nil
+			}
+		}
+	}
+
+	return 0, fmt.Errorf("no winners")
+}
+
+const BINGO_BOARD_SIZE = 5
+
+type bingoBoard struct {
+	squares [][]*bingoSquare
+}
+
+func (t bingoBoard) String() string {
+	var builder strings.Builder
+	for _, row := range t.squares {
+		for _, square := range row {
+			builder.WriteString(fmt.Sprint(square))
+		}
+		builder.WriteString("\n")
+	}
+	return builder.String()
+}
+
+type bingoSquare struct {
+	number int
+	marked bool
+}
+
+func (t bingoSquare) String() string {
+	if t.marked {
+		return fmt.Sprintf(" _%2v_ ", t.number)
+	} else {
+		return fmt.Sprintf("  %2v  ", t.number)
+	}
+}
+
+func (board *bingoBoard) mark(number int) bool {
+	for _, row := range board.squares {
+		for _, square := range row {
+			if square.number == number {
+				square.marked = true
+			}
+		}
+	}
+	return board.isWinner()
+}
+
+func (board *bingoBoard) isWinner() bool {
+	return board.hasMarkedRow() || board.hasMarkedColumn()
+}
+
+func (board *bingoBoard) hasMarkedRow() bool {
+OUTER:
+	for _, row := range board.squares {
+		for _, square := range row {
+			if !square.marked {
+				continue OUTER
+			}
+		}
+		return true
+	}
+
+	return false
+}
+
+func (board *bingoBoard) hasMarkedColumn() bool {
+OUTER:
+	for col := 0; col < BINGO_BOARD_SIZE; col++ {
+		for row := 0; row < BINGO_BOARD_SIZE; row++ {
+			square := board.squares[row][col]
+			if !square.marked {
+				continue OUTER
+			}
+		}
+		return true
+	}
+
+	return false
+}
+
+func (board *bingoBoard) score() int {
+	score := 0
+	for _, row := range board.squares {
+		for _, square := range row {
+			if !square.marked {
+				score += square.number
+			}
+		}
+	}
+	return score
+}
+
+func parseBoard(lines []string) (*bingoBoard, error) {
+	if len(lines) != BINGO_BOARD_SIZE {
+		return nil, fmt.Errorf("boards want %v lines", BINGO_BOARD_SIZE)
+	}
+
+	squares := make([][]*bingoSquare, BINGO_BOARD_SIZE)
+	for i, line := range lines {
+		nums, err := parseNumbers(strings.Fields(line))
+		if err != nil {
+			return nil, err
+		}
+		squares[i] = make([]*bingoSquare, BINGO_BOARD_SIZE)
```

Comment:
is this needed, or will the `make([][]...)` on L195 handle it?

Reply:
nope, elements default to `nil`

Reply:
but if I use an [array](https://go.dev/ref/spec#Array_types) instead of a slice, I get initialization for free